File: src/main/java/com/yai/StatementEvaluator.java
package com.yai;

import org.antlr.v4.runtime.tree.ParseTree;

import java.util.LinkedHashMap;
import java.util.Map;

public class StatementEvaluator extends YaInterpreterBaseVisitor<Map<String, Double>> {

    private final YaInterpreterLexer lexer;
    private final Map<String, Double> variables = new LinkedHashMap<>();

    private String currentVariable = null;

    public StatementEvaluator(YaInterpreterLexer lexer, ParseTree tree) {
        this.lexer = lexer;
        visit(tree);
    }

    public Map<String, Double> getVariables() {
        return variables;
    }


    public Map<String, Double> visitStatement(YaInterpreterParser.StatementContext ctx) {
        for (int i = 0; i < ctx.expression().size(); i++) {
            visit(ctx.variable(i));
            double result = new ExpressionEvaluator(lexer, variables, ctx.expression(i)).getResult();
            variables.put(currentVariable, result);
        }
        return variables;
    }

    @Override
    public Map<String, Double> visitVariable(YaInterpreterParser.VariableContext ctx) {
        currentVariable = ctx.getText();
        return variables;
    }
}

File: src/main/kotlin/com/yai/StatementEvaluator.kt
//package com.yai
//
//import com.yai.YaInterpreterParser.StatementContext
//import com.yai.YaInterpreterParser.VariableContext
//import org.antlr.v4.runtime.tree.ParseTree
//
//class StatementEvaluator(
//    private val lexer: YaInterpreterLexer,
//    tree: ParseTree?
//) : YaInterpreterBaseVisitor<Map<String, Double>>() {
//
//    private val variables: MutableMap<String, Double> = LinkedHashMap()
//    private var currentVariable: String = ""
//
//    init {
//        visit(tree)
//    }
//
//    fun getVariables(): Map<String, Double> {
//        return variables
//    }
//
//    override fun visitStatement(ctx: StatementContext?): Map<String, Double> {
//        for (i in ctx!!.expression().indices) {
//            visit(ctx.variable(i) as ParseTree)
//            val result = ExpressionEvaluator(lexer, variables, ctx.expression(i) as ParseTree).result as Double
//            variables[currentVariable] = result
//        }
//        return variables
//    }
//
//    override fun visitVariable(ctx: VariableContext?): Map<String, Double> {
//        currentVariable = ctx!!.text
//        return variables
//    }
//
//}